Add support to log first exception in flush() at warning-level

dzou opened the pull request:
This makes a change in `BoundedAsyncReporter` to log the first exception thrown as a warning with a "check more at FINE" message; subsequent exceptions are logged to FINE level.

Fixes #154.

cc/ @meltsufin @elefeint

shakuzen commented on core/src/main/java/zipkin2/reporter/AsyncReporter.java:
```diff
@@ -285,11 +289,22 @@ void flush(BufferNextMessage<S> bundler) {
+        if (shouldWarnException) {
+          logger.log(WARNING, "Spans were dropped due to exceptions. "
+            + "All subsequent errors will be logged at FINE level.");
+          logLevel = WARNING;
```
Won't this make subsequent logs at `WARNING` level?

shakuzen replied:
Sorry, I see what this is doing now. Nevermind my comment then.

adriancole commented:
something like this. note you may need to rejig the flusher or something to
allow control over the logger similarly how done here

https://github.com/openzipkin/zipkin/blob/master/zipkin/src/test/java/zipkin2/internal/SpanNodeTest.java#L259

On Thu, Aug 1, 2019 at 10:31 AM Tommy Ludwig <notifications@github.com>
wrote:

> *@shakuzen* commented on this pull request.
>
> I'm not sure where we have a logging test to base one off of without
> searching more. I'll search more later if @adriancole
> <https://github.com/adriancole> doesn't know off the top of his head
> whereabouts one would be.
> ------------------------------
>
> In core/src/main/java/zipkin2/reporter/AsyncReporter.java
> <https://github.com/openzipkin/zipkin-reporter-java/pull/155#discussion_r309500926>
> :
>
> > @@ -285,11 +289,22 @@ void flush(BufferNextMessage<S> bundler) {
>          Call.propagateIfFatal(t);
>          metrics.incrementMessagesDropped(t);
>          metrics.incrementSpansDropped(count);
> -        if (logger.isLoggable(FINE)) {
> -          logger.log(FINE,
> -              format("Dropped %s spans due to %s(%s)", count, t.getClass().getSimpleName(),
> -                  t.getMessage() == null ? "" : t.getMessage()), t);
> +
> +        Level logLevel = FINE;
> +
> +        if (shouldWarnException) {
> +          logger.log(WARNING, "Spans were dropped due to exceptions. "
> +            + "All subsequent errors will be logged at FINE level.");
> +          logLevel = WARNING;
>
> Won't this make subsequent logs at WARNING level?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/zipkin-reporter-java/pull/155?email_source=notifications&email_token=AAAPVV447JCYCEB5OWNXPCTQCJDHXA5CNFSM4IIK4DXKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCAHDYOQ#pullrequestreview-269368378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPVVZDDVSL3AT6M6IFGG3QCJDHXANCNFSM4IIK4DXA>
> .
>

dzou commented:
Thanks for the tips. Added unit test for this.

adriancole commented:
I changed some config to make something blow up and it works as expected:

```
2019-08-09 12:12:11.992  WARN 2638 --- [orageComponent}] z.r.AsyncReporter$BoundedAsyncReporter   : Spans were dropped due to exceptions. All subsequent errors will be logged at FINE level.
2019-08-09 12:12:11.993  WARN 2638 --- [orageComponent}] z.r.AsyncReporter$BoundedAsyncReporter   : Dropped 16 spans due to RejectedExecutionException(rejected execution of processing of [11076][indices:data/write/bulk[s][p]]: request: BulkShardRequest [[zipkin:span-2019-08-09][0]] containing [3] requests, target allocation id: DDNWNWk6S8yUbOiJgDnFLQ, primary term: 1 on EsThreadPoolExecutor[name = kZ5Vr2U/write, queue capacity = 10, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@34d5c81f[Running, pool size = 8, active threads = 2, queued tasks = 10, completed tasks = 7322]])

java.util.concurrent.RejectedExecutionException: rejected execution of processing of [11076][indices:data/write/bulk[s][p]]: request: BulkShardRequest [[zipkin:span-2019-08-09][0]] containing [3] requests, target allocation id: DDNWNWk6S8yUbOiJgDnFLQ, primary term: 1 on EsThreadPoolExecutor[name = kZ5Vr2U/write, queue capacity = 10, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@34d5c81f[Running, pool size = 8, active threads = 2, queued tasks = 10, completed tasks = 7322]]
	at zipkin2.elasticsearch.internal.BulkCallBuilder$1.convert(BulkCallBuilder.java:62) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
	at zipkin2.elasticsearch.internal.BulkCallBuilder$1.convert(BulkCallBuilder.java:50) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
	at zipkin2.elasticsearch.internal.client.HttpCall.parseResponse(HttpCall.java:217) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
	at zipkin2.elasticsearch.internal.client.HttpCall.lambda$doEnqueue$0(HttpCall.java:128) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
	at com.linecorp.armeria.common.HttpMessageAggregator.onComplete(HttpMessageAggregator.java:98) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.FilteredStreamMessage$FilteringSubscriber.onComplete(FilteredStreamMessage.java:290) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:363) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:194) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:363) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:306) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:253) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.addObjectOrEvent(DefaultStreamMessage.java:239) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.common.stream.DefaultStreamMessage.close(DefaultStreamMessage.java:369) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.closeAction(HttpResponseDecoder.java:304) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:287) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:275) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at com.linecorp.armeria.client.Http1ResponseDecoder.channelRead(Http1ResponseDecoder.java:212) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328) ~[netty-codec-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302) ~[netty-codec-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:154) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:697) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
	at java.lang.Thread.run(Thread.java:834) [?:?]

```

dzou commented:
> I changed some config to make something blow up and it works as expected:
> 
> ```
> 2019-08-09 12:12:11.992  WARN 2638 --- [orageComponent}] z.r.AsyncReporter$BoundedAsyncReporter   : Spans were dropped due to exceptions. All subsequent errors will be logged at FINE level.
> 2019-08-09 12:12:11.993  WARN 2638 --- [orageComponent}] z.r.AsyncReporter$BoundedAsyncReporter   : Dropped 16 spans due to RejectedExecutionException(rejected execution of processing of [11076][indices:data/write/bulk[s][p]]: request: BulkShardRequest [[zipkin:span-2019-08-09][0]] containing [3] requests, target allocation id: DDNWNWk6S8yUbOiJgDnFLQ, primary term: 1 on EsThreadPoolExecutor[name = kZ5Vr2U/write, queue capacity = 10, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@34d5c81f[Running, pool size = 8, active threads = 2, queued tasks = 10, completed tasks = 7322]])
> 
> java.util.concurrent.RejectedExecutionException: rejected execution of processing of [11076][indices:data/write/bulk[s][p]]: request: BulkShardRequest [[zipkin:span-2019-08-09][0]] containing [3] requests, target allocation id: DDNWNWk6S8yUbOiJgDnFLQ, primary term: 1 on EsThreadPoolExecutor[name = kZ5Vr2U/write, queue capacity = 10, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@34d5c81f[Running, pool size = 8, active threads = 2, queued tasks = 10, completed tasks = 7322]]
> 	at zipkin2.elasticsearch.internal.BulkCallBuilder$1.convert(BulkCallBuilder.java:62) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
> 	at zipkin2.elasticsearch.internal.BulkCallBuilder$1.convert(BulkCallBuilder.java:50) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
> 	at zipkin2.elasticsearch.internal.client.HttpCall.parseResponse(HttpCall.java:217) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
> 	at zipkin2.elasticsearch.internal.client.HttpCall.lambda$doEnqueue$0(HttpCall.java:128) ~[zipkin-storage-elasticsearch-2.15.1-SNAPSHOT.jar!/:?]
> 	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930) ~[?:?]
> 	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907) ~[?:?]
> 	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
> 	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
> 	at com.linecorp.armeria.common.HttpMessageAggregator.onComplete(HttpMessageAggregator.java:98) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.FilteredStreamMessage$FilteringSubscriber.onComplete(FilteredStreamMessage.java:290) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.AbstractStreamMessage$CloseEvent.notifySubscriber(AbstractStreamMessage.java:363) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberOfCloseEvent(DefaultStreamMessage.java:194) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.handleCloseEvent(DefaultStreamMessage.java:363) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber0(DefaultStreamMessage.java:306) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:253) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.addObjectOrEvent(DefaultStreamMessage.java:239) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.common.stream.DefaultStreamMessage.close(DefaultStreamMessage.java:369) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.closeAction(HttpResponseDecoder.java:304) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:287) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.client.HttpResponseDecoder$HttpResponseWrapper.close(HttpResponseDecoder.java:275) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at com.linecorp.armeria.client.Http1ResponseDecoder.channelRead(Http1ResponseDecoder.java:212) ~[armeria-master-armeria-0.27.0.Final-g91bbcf2-1258.jar!/:?]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328) ~[netty-codec-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302) ~[netty-codec-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:154) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287) ~[netty-handler-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:697) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511) ~[netty-transport-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.38.Final.jar!/:4.1.38.Final]
> 	at java.lang.Thread.run(Thread.java:834) [?:?]
> ```

Sweeet -- seems helpful!
